flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStorage.java

```diff
@@ -107,6 +105,13 @@ public CheckpointStorageLocation initializeLocationForCheckpoint(long checkpoint
 		// prepare all the paths needed for the checkpoints
 		final Path checkpointDir = createCheckpointDirectory(checkpointsDirectory, checkpointId);
 
+		if (!areDirectoriesCreated) {
+			fileSystem.mkdirs(checkpointsDirectory);
+			fileSystem.mkdirs(sharedStateDirectory);
```

Comment:
Now that only shared and taskowned would be left, I wonder if it would not be cleaner to just create them in an init method that is only called on the JM/checkpoint coordinator side. Do you think there is still benefit in true lazyness, like only create the dirs when the first checkpoint is actually triggered. Or is it sufficient if not all TMs will also make those calls? From code simplicity, it would be better if we can avoid any lazy initialization flags.

Reply:
At first, I planed to add a new initialization method to create those directories. However, I found current `initializeLocationForCheckpoint` method already has some initialization semantics. I think since all checkpoints would first start from checkpoint coordinator, and no clear annotation for `CheckpointStorage` I'm not sure adding a new method would whether break the compatibility.

For open-close-principle, I think it would be better to just adding an initialization method. And to be honest, I would not find any benefits in lazy initialization.

If adding a new method, do you have any suggestion about the method name to distinguish from current `initializeLocationForCheckpoint`?

Reply:
I think it would be reasonable to add a new method. As for the names, maybe something like a `initializeLocation` and `initializeLocationForCheckpoint` -- meaning (as we can explain in the docs):  one method initializes that initializes the location _in general_ and one method that initializes things for a specific checkpoint x. So shared directories etc can go to general init, and the checkpoint specific folder goes to the existing method.

Reply:
Furthermore, I think we could also consider splitting the interface into one that is JM side, owned by checkpoint coordinator and has the init lifecycle methods, and one that only has the methods relevant for TM. Implementations inherit both, but his could shield TMs from accidentally calling init methods. Would that make sense?

Reply:
I think a new interface `ChekcpointCoordinatorStorgae` extending `ChckecpointStorgae` with `#supportsHighlyAvailableStorage` ,  `#hasDefaultSavepointLocation`, `#resolveCheckpoint`, `#initializeLocationForCheckpoint` and `#initializeLocationForSavepoint` methods could be introduced, while previous `CheckpointStorage` would only have two methods: `#resolveCheckpointStorageLocation` and `#createTaskOwnedStateStream`.

In JM side, we could cast the created storage as `ChekcpointCoordinatorStorgae` so that we could have those methods while TM side would not call those methods accidentally.

I think this would be clearer for users to distinguish checkpoint storage on different sides. However, I think this is more like a interface refactor which contains more than just "avoid to send mkdirs from task sides". Shall we also include this refactor in this PR?

Reply:
Yes, that sounds like a good split and would exactly help with that problem.
The best way to handle this PR-wise would be:
- create a JIRA for splitting the interface into something like and JM/TM side.
- open PR for that change.
- have this PR rebased, based on the interface split-PR.

We should keep the cost/benefit ratio in mind. If you think rebasing would take you a lot of time (I think it should not) then just do it in this PR. Otherwise, I suggest the above strategy.

Reply:
Sounds good and thanks for your review and suggestions, already created the related JIRA [FLINK-11874](https://issues.apache.org/jira/browse/FLINK-11874)